deploy/runtools/firesim_topology_elements.py

```diff
@@ -474,9 +474,37 @@ def get_sim_kill_command(self, slotno: int) -> str:
         """
         return self.get_resolved_server_hardware_config().get_kill_simulation_command()
 
+    def get_tarball_files_paths(self) -> List[Tuple[str, str]]:
+        """ Return local and remote paths of all stuff destined for the driver tarball 
+        The returned paths in the tuple are [local_path, remote_path]. """
+        all_paths = []
+
+        driver_path = self.get_resolved_server_hardware_config().get_local_driver_path()
+        all_paths.append((driver_path, ''))
+        all_paths.append((self.get_resolved_server_hardware_config().get_local_runtime_conf_path(), ''))
+
+        # shared libraries
+        all_paths += get_local_shared_libraries(driver_path)
+        all_paths += self.get_resolved_server_hardware_config().get_additional_required_sim_files()
+
+        all_paths += self.get_job().get_siminputs()
+        return all_paths
+    
+    def get_tarball_path_pair(self) -> List[Tuple[str, str]]:
+        """ Return local and remote paths of the actual driver tarball, not files inside"""
+        all_paths = []
+
+        if not isinstance(self.server_hardware_config, str) and self.server_hardware_config is not None and self.server_hardware_config.driver_tar_uri is not None:
```

Comment:
This might be a question for @abejgonzalez , but can you explain is this `isinstance(self.server_hardware_config, str)` here.  Under what conditions can it be a string?

Reply:
In the user topologies, you can assign each server node of a simulation to a specific HW config by giving a string (https://github.com/firesim/firesim/blob/7aa7ee22f0ab91d29340d1842f34acb7952c20bc/deploy/runtools/user_topology.py#L407-L414). This was an old API to let people do heterogeneous simulations with different HW cfgs.